File: src/main/java/org/chombo/mr/SeasonalDetector.java
/*
 * chombo: Hadoop Map Reduce utility
 * Author: Pranab Ghosh
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */


package org.chombo.mr;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.chombo.util.BasicUtils;
import org.chombo.util.Pair;
import org.chombo.util.SeasonalAnalyzer;
import org.chombo.util.Tuple;
import org.chombo.util.Utility;

/**
 * Detects seasonality by building histogram based on specified type of sesonality
 * @author pranab
 *
 */
public class SeasonalDetector  extends Configured implements Tool {
	private static String configDelim = ",";
    private static final String AGGR_COUNT = "count";
    private static final String AGGR_SUM = "sum";
    private static final String AGGR_AVG = "average";

	@Override
	public int run(String[] args) throws Exception {
        Job job = new Job(getConf());
        String jobName = "Seasonal aggregator  for numerical attributes";
        job.setJobName(jobName);
        
        job.setJarByClass(SeasonalDetector.class);

        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileOutputFormat.setOutputPath(job, new Path(args[1]));

        Utility.setConfiguration(job.getConfiguration(), "chombo");
        job.setMapperClass(SeasonalDetector.AggregatorMapper.class);
        job.setReducerClass(SeasonalDetector.AggregateReducer.class);
        job.setCombinerClass(SeasonalDetector.AggregateCombiner.class);
        
        job.setMapOutputKeyClass(Tuple.class);
        job.setMapOutputValueClass(Tuple.class);

        job.setOutputKeyClass(NullWritable.class);
        job.setOutputValueClass(Text.class);

        int numReducer = job.getConfiguration().getInt("sed.num.reducer", -1);
        numReducer = -1 == numReducer ? job.getConfiguration().getInt("num.reducer", 1) : numReducer;
        job.setNumReduceTasks(numReducer);

        int status =  job.waitForCompletion(true) ? 0 : 1;
        return status;
	}

	/**
	 * @author pranab
	 *
	 */
	public static class AggregatorMapper extends Mapper<LongWritable, Text, Tuple, Tuple> {
		private Tuple outKey = new Tuple();
		private Tuple outVal = new Tuple();
		private int[]  attributes;
        private String fieldDelimRegex;
        private String[] items;
        private int[] idOrdinals;
        private int timeStampFieldOrdinal;
        private String seasonalCycleType;
        private int cycleIndex;
        private long timeStamp;
        private String aggregatorType;
        private SeasonalAnalyzer seasonalAnalyzer;
        
        protected void setup(Context context) throws IOException, InterruptedException {
        	Configuration config = context.getConfiguration();
        	fieldDelimRegex = config.get("field.delim.regex", ",");
        	attributes = Utility.assertIntArrayConfigParam(config, "sed.quant.attr.list", fieldDelimRegex, 
        			"missing quant attribute list");
        	idOrdinals = Utility.intArrayFromString(config.get("sed.id.field.ordinals"), configDelim);
        	timeStampFieldOrdinal = Utility.assertIntConfigParam(config,"sed.time.stamp.field.ordinal", 
        			"missing timestamp field ordinal");
        	seasonalCycleType = Utility.assertStringConfigParam(config,"sed.seasonal.cycle.type", 
        			"missing seasonal cycle type");
        	aggregatorType = config.get("sed.aggregator.type", AGGR_AVG);
        	
    		seasonalAnalyzer = new SeasonalAnalyzer(seasonalCycleType);

    		//additional configuration
        	if (seasonalCycleType.equals(SeasonalAnalyzer.HOUR_RANGE_OF_WEEK_DAY ) ||  
        			seasonalCycleType.equals(SeasonalAnalyzer.HOUR_RANGE_OF_WEEK_END_DAY ) ) {
        		Map<Integer, Integer>  hourRanges = Utility.assertIntegerIntegerMapConfigParam(config, "sed.hour.groups", 
        				Utility.configDelim, Utility.configSubFieldDelim, "missing hour groups", true);
        		seasonalAnalyzer.setHourRanges(hourRanges);
        	} else if (seasonalCycleType.equals(SeasonalAnalyzer.ANY_TIME_RANGE)) {
        		List<Pair<Integer, Integer>> timeRanges = Utility.assertIntPairListConfigParam(config, "sed.time.ranges", 
        				Utility.configDelim, Utility.configSubFieldDelim, "missing hour groups");
        		seasonalAnalyzer.setTimeRanges(timeRanges);
        	}
        	
        	//time zone adjustment
        	int  timeZoneShiftHours = config.getInt("sed.time.zone.shift.hours",  0);
        	if (timeZoneShiftHours > 0) {
        		seasonalAnalyzer.setTimeZoneShiftHours(timeZoneShiftHours);
        	}
        	
        	//timestamp unit
        	boolean timeStampInMili = config.getBoolean("sed.time.stamp.in.mili", true);
        	seasonalAnalyzer.setTimeStampInMili(timeStampInMili);
       }
		
        @Override
        protected void map(LongWritable key, Text value, Context context)
            throws IOException, InterruptedException {
            items  =  value.toString().split(fieldDelimRegex, -1);
            
            timeStamp = Long.parseLong(items[timeStampFieldOrdinal]);
            cycleIndex = seasonalAnalyzer.getCycleIndex(timeStamp);
            
            if (cycleIndex >= 0) {
	        	for (int attr : attributes) {
	            	outKey.initialize();
	            	outVal.initialize();
	            	
	            	if (null != idOrdinals) {
	            		outKey.addFromArray(items, idOrdinals);
	            	}
	            	outKey.add(seasonalAnalyzer.getParentCycleIndex(), cycleIndex, attr);
	            	
	            	if (aggregatorType.equals(AGGR_COUNT)) {
	            		outVal.add(1);
	            	} else if (aggregatorType.equals(AGGR_SUM)) {
	            		outVal.add(Double.parseDouble(items[attr]));
	            	} else if (aggregatorType.equals(AGGR_AVG)) {
	            		outVal.add(1, Double.parseDouble(items[attr]));
	            	} else {
	        			throw new IllegalArgumentException("invalid aggregation function");
	        		}
	            	context.write(outKey, outVal);
	        	}
            }
        }       
	}
	
	/**
	 * @author pranab
	 *
	 */
	public static class AggregateCombiner extends Reducer<Tuple, Tuple, Tuple, Tuple> {
		private Tuple outVal = new Tuple();
		private int totalCount;
		private double sum;
		private String aggregatorType;

        /* (non-Javadoc)
         * @see org.apache.hadoop.mapreduce.Reducer#setup(org.apache.hadoop.mapreduce.Reducer.Context)
         */
        protected void setup(Context context) throws IOException, InterruptedException {
        	Configuration config = context.getConfiguration();
        	aggregatorType = config.get("sed.aggregator.type");
        }
        
        /* (non-Javadoc)
         * @see org.apache.hadoop.mapreduce.Reducer#reduce(KEYIN, java.lang.Iterable, org.apache.hadoop.mapreduce.Reducer.Context)
         */
        protected void reduce(Tuple  key, Iterable<Tuple> values, Context context)
        		throws IOException, InterruptedException {
    		sum = 0;
    		totalCount = 0;
    		for (Tuple val : values) {
            	if (aggregatorType.equals(AGGR_COUNT)) {
            		totalCount += val.getInt(0);
            	} else if (aggregatorType.equals(AGGR_SUM)) {
            		sum  += val.getDouble(0);
            	} else if (aggregatorType.equals(AGGR_AVG)) {
            		totalCount += val.getInt(0);
            		sum  += val.getDouble(1);
            	}
    		}
        
    		outVal.initialize();
    		if (aggregatorType.equals(AGGR_COUNT)) {
    			outVal.add(totalCount);
    		} else if (aggregatorType.equals(AGGR_SUM)) {
    			outVal.add(sum);
    		} else if (aggregatorType.equals(AGGR_AVG)) {
    			outVal.add(totalCount, sum);
    		} 
        	context.write(key, outVal);       	
        }
	}
	
    /**
    * @author pranab
    *
    */
    public static class  AggregateReducer extends Reducer<Tuple, Tuple, NullWritable, Text> {
    	private Text outVal = new Text();
 		private StringBuilder stBld =  new StringBuilder();;
		private String fieldDelim;
		private double sum;
		private int totalCount;
		private double average;
		private String aggregatorType;
		private int outputPrecision;
		
		/* (non-Javadoc)
		 * @see org.apache.hadoop.mapreduce.Reducer#setup(org.apache.hadoop.mapreduce.Reducer.Context)
		 */
		protected void setup(Context context) throws IOException, InterruptedException {
        	Configuration config = context.getConfiguration();
        	fieldDelim = config.get("field.delim.out", ",");
           	aggregatorType = config.get("sed.aggregator.type");
           	outputPrecision = config.getInt("sed.output.precision", 3);
 		}

        /* (non-Javadoc)
         * @see org.apache.hadoop.mapreduce.Reducer#reduce(KEYIN, java.lang.Iterable, org.apache.hadoop.mapreduce.Reducer.Context)
         */
        protected void reduce(Tuple  key, Iterable<Tuple> values, Context context)
        		throws IOException, InterruptedException {
        	key.setDelim(fieldDelim);
    		sum = 0;
    		totalCount = 0;
    		for (Tuple val : values) {
            	if (aggregatorType.equals(AGGR_COUNT)) {
            		totalCount += val.getInt(0);
            	} else if (aggregatorType.equals(AGGR_SUM)) {
            		sum  += val.getDouble(0);
            	}  else if (aggregatorType.equals(AGGR_AVG)) {
            		totalCount += val.getInt(0);
            		sum  += val.getDouble(1);
            	}
    		}
    		
        	if (aggregatorType.equals(AGGR_COUNT)) {
        		outVal.set(key.toString() + fieldDelim +  totalCount);
        	} else if (aggregatorType.equals(AGGR_SUM)) {
        		 setOutput(key, sum); 
        	} else if (aggregatorType.equals(AGGR_AVG)) {
        		average = sum / totalCount;
          		 setOutput(key, average); 
            }
        	context.write(NullWritable.get(), outVal);
        }		
        
        /**
         * @param key
         * @param value
         */
        private void setOutput(Tuple  key, double value) {
    		if (0 == outputPrecision) {
        		outVal.set(key.toString() +fieldDelim  +(int)value);
    		} else {
        		outVal.set(key.toString() +fieldDelim  +BasicUtils.formatDouble(value, outputPrecision));
    		}
        }
 	}
 
	/**
	 * @param args
	 */
	public static void main(String[] args) throws Exception {
		int exitCode = ToolRunner.run(new SeasonalDetector(), args);
		System.exit(exitCode);
	}
 
}

File: spark/src/main/scala/org/chombo/spark/explore/SeasonalDetector.scala
/*
 * chombo-spark: etl on spark
 * Author: Pranab Ghosh
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package org.chombo.spark.explore

import org.chombo.spark.common.JobConfiguration
import org.apache.spark.SparkContext
import scala.collection.JavaConverters._
import org.chombo.util.BasicUtils
import org.chombo.util.SeasonalAnalyzer
import com.typesafe.config.Config
import org.chombo.spark.common.Record

object SeasonalDetector extends JobConfiguration {
   
  /**
   * @param args
   * @return
   */
   def main(args: Array[String]) {
	   val appName = "seasonalDetector"
	   val Array(inputPath: String, outputPath: String, configFile: String) = getCommandLineArgs(args, 3)
	   val config = createConfig(configFile)
	   val sparkConf = createSparkConf(appName, config, false)
	   val sparkCntxt = new SparkContext(sparkConf)
	   val appConfig = config.getConfig(appName)
	   
	   //configuration params
	   val fieldDelimIn = appConfig.getString("field.delim.in")
	   val fieldDelimOut = appConfig.getString("field.delim.out")
	   val attributes = getMandatoryIntListParam(appConfig, "attributes").asScala.toArray
	   val idOrdinals = getOptionalIntListParam(appConfig, "id.ordinals")
	   val timeStampFieldOrdinal = getMandatoryIntParam(appConfig, "timeStamp.field.ordinal", 
	       "missing time stamp field ordinal configuration")
	   val seasonalCycleType = getMandatoryStringParam(appConfig, "seasonal.cycle.type", 
	       "missing seasonal cycle type configuration")
	   val seasonalAnalyzer = new SeasonalAnalyzer(seasonalCycleType)
	   
	   //additional configuration
	   seasonalCycleType match {
	     case SeasonalAnalyzer.HOUR_RANGE_OF_WEEK_DAY => seasonalAnalyzer.setHourRanges(getHourGroups(appConfig))
	     case SeasonalAnalyzer.HOUR_RANGE_OF_WEEK_END_DAY => seasonalAnalyzer.setHourRanges(getHourGroups(appConfig))
	     case SeasonalAnalyzer.ANY_TIME_RANGE => seasonalAnalyzer.setTimeRanges(getTimeRange(appConfig))
	   }
	   
	   val timeZoneShiftHours = this.getIntParamOrElse(appConfig, "time.zone.shift.hours", 0)
	   if (timeZoneShiftHours > 0) seasonalAnalyzer.setTimeZoneShiftHours(timeZoneShiftHours)
       val timeStampInMili = this.getBooleanParamOrElse(appConfig, "time.stamp.in.mili", true)
	   seasonalAnalyzer.setTimeStampInMili(timeStampInMili)
	   val aggregatorType = this.getStringParamOrElse(appConfig, "aggregator.type", "average")
	   
	   val debugOn = appConfig.getBoolean("debug.on")
	   val saveOutput = appConfig.getBoolean("save.output")
	   
	   val data = sparkCntxt.textFile(inputPath)
	   
	   //key by parent cycle index, cycle index, attribute
	   val cycleKeyedRecs = data.flatMap(line => {
	     val items  =  line.split(fieldDelimIn, -1) 
	     val lineRec = Record(items)
	     val timeStamp = BasicUtils.getLongField(items, timeStampFieldOrdinal)
	     val cycleIndex = seasonalAnalyzer.getCycleIndex(timeStamp)
	     val validCycleIndex = cycleIndex >= 0
	     val cycleKeyedRecs = validCycleIndex match {
	       case true => {
	         val keyValRecs = attributes.map(a => {
	           val key = idOrdinals match {
	             case Some(idOrds : java.util.List[Integer]) => {
	               val key = Record(idOrds.size() + 3)
	               key.add(lineRec, idOrds).add(seasonalAnalyzer.getParentCycleIndex(), cycleIndex, a)
	               key
	             }
	             case None => {
	               val key = Record(3)
	               key.add(seasonalAnalyzer.getParentCycleIndex(), cycleIndex, a)
	               key
	             }
	           }
	           
	           val value = Record(1)
	           value.add(1, BasicUtils.getDoubleField(items, a))
	           (key, value)
	         })
	         keyValRecs
	       }
	       case false => {
	         Array[(org.chombo.spark.common.Record, org.chombo.spark.common.Record)]()
	       }
	     }
	   cycleKeyedRecs
	   })

	   //reduce
	   val redCycleKeyedRecs = cycleKeyedRecs.reduceByKey((v1, v2) => {
	     val count = v1.getInt(0) + v2.getInt(0)
	     val sum = v1.getDouble(1) + v2.getDouble(1)
	     val aggrRec = Record(2)
	     aggrRec.add(count, sum)
	     aggrRec
	   })
	   
	   //aggregate
	   val aggrCycleKeyedRecs = redCycleKeyedRecs.mapValues(v => {
	     val rec = Record(1)
	     aggregatorType match {
	       case "count" => rec.add(v.getInt(0))
	       case "sum" => rec.add(v.getDouble(1))
	       case "average" => rec.add(v.getDouble(1) / v.getInt(0))
	     }
	     rec
	   })
	   
       if (debugOn) {
         val records = aggrCycleKeyedRecs.collect
         records.foreach(r => println(r._1.toString() + fieldDelimOut + r._2.toString()))
       }
	   
	   if(saveOutput) {	   
	     aggrCycleKeyedRecs.saveAsTextFile(outputPath) 
	   }
	   
	   
   }
   
   
   /**
   * @param appConfig
   * @return
   */
   def getHourGroups(appConfig : Config) : java.util.Map[Integer, Integer] = {
     val hourGroupsStr = getMandatoryStringParam(appConfig, "hour.groups", "missing seasonal hout group configuration")
     BasicUtils.integerIntegerMapFromString(hourGroupsStr, BasicUtils.DEF_FIELD_DELIM, BasicUtils.DEF_SUB_FIELD_DELIM, true)
   }
   
   /**
   * @param appConfig
   * @return
   */
   def getTimeRange(appConfig : Config) :  java.util.List[org.chombo.util.Pair[Integer,Integer]]  = {
     val timeRangeStr = getMandatoryStringParam(appConfig, "time.range", "missing seasonal time range configuration")
     BasicUtils.getIntPairList(timeRangeStr, BasicUtils.DEF_FIELD_DELIM, BasicUtils.DEF_SUB_FIELD_DELIM)
   }
}